x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java

```diff
@@ -414,7 +415,12 @@ public ClusterState execute(ClusterState currentState) {
 
                     if (areClusterStatesCompatibleForRebalance(clusterState, currentState)) {
                         isUpdated = true;
-                        return update(currentState, rebalancedMetadata);
+                        ClusterState updatedState = update(currentState, rebalancedMetadata);
+                        if (TrainedModelAssignmentMetadata.fromState(currentState)
```

Comment:
Can the `update()` method return a `Tuple<Boolean, ClusterState>` where the boolean indicates a change. It saves extracting the metadata from cluster state again and the `equals()` check